_posts/2019-12-25-technical-debt.md

```diff
@@ -0,0 +1,107 @@
+
+Furthermore, this brings to mind how we would utilise these metrics in a system that will continue to be used in the forseeable future but isn't undergoing rapid change and remains stable. 
+
+- Lead Time For Changes would trend towards zero; no changes means no lead time.
+- Change Failure Rate would trend towards zero; no changes mean that no changes can fail.
+- Deployment Frequency would trend towards zero; no changes (often) mean no deployments.
+- Time to Restore Service; assuming system is indeed stable (humor this), this would also trend to zero as the system would never excercise the need to be restored.
+
+A perfect system would therefore trend to zero on all measures but we know such a system does not exist; all systems fail at some point, resulting in a need to restore a service. This would likely result in a change and therefore effect all the other metrics. This does mean that systems that are not underdoing rapid change (or failing often) can harbour serious technical debt. There is an inherent risk that a major outage could occur that is difficult to repair. In younger systems deployment is an inherently chaotic event that affords the opportunity to impact all the metrics and therefore discover opportunity for improvement. This opportunity is not as actively present in maturer systems and has to be introduced. Chaos engineering therefore becomes more important as systems approach a certain age and criticality to the business; it introduces a kind of controlled change that normally only occurs during active development. In saying this it may be appropriate for teams starting out with chaos engineering to regularly schedule the deployment of code to actively test the ability to deploy frequently even if no changes have been made.
```

Comment:
"In younger systems deployment..."  insert a comma between systems and deployment

Reply:
Also don't understand this sentence. What do "younger systems" have to do with "inherently chaotic" and "impacting all metrics" ? You mean the process is not stable? (If you want a side track, see Deming on Special Cause Variation and Common Cause Variation)

Reply:
This whole para starts well but trails off into chaos... pun intended. What's the point?

Reply:
It comes something I've noticed with systems that are no are longer receiving active development but still need to be kept up. The broker proxy being an example of this.

- Development finishes
- Thing runs for a long time with no-one touching it
- There's a bug, and something needs to be modified in an emergency.
- Someone tries to make a change, but the pipeline for whatever reason is now busted.

Basically, you need to constantly introducing some level of change to exercise the system - it doesn't have to be a code level change - there's a lot on inherent 'chaos' in the average deployment (standing up, tearing down, switching traffic, DNS changes) to uncover this without anything additional.

Going a step further - you can induce errors in a controlled manner to increase the resiliency and uncover debt you weren't even aware of.

Reply:
Basically two goals;

- Reduce Mean-Time-To-Detection, and,
- Hopefully uncover smaller errors that can be handled ASAP, rather than multiple ones that may have accrued over time.